plugins/tedge_apama_plugin/src/main.rs

```diff
@@ -48,128 +48,254 @@ pub enum PluginOp {
+    match module.rsplit_once("::") {
+        Some((prefix, suffix)) => match suffix {
+            PROJECT_SUFFIX => Ok(ApamaModule::Project(prefix.into())),
+            MON_SUFFIX => Ok(ApamaModule::MonFile(prefix.into())),
+            unsupported_type => Err(InternalError::UnsupportedModuleType {
+                module_type: unsupported_type.into(),
+            }),
+        },
+        None => Err(InternalError::ModuleTypeNotProvided {
+            module_name: module.into(),
+        }),
+    }
+}
 
-            println!(
-                "Installing newly extracted project to {}",
-                TEDGE_APAMA_PROJECT_DIR
-            );
-            fs::create_dir_all(tedge_apama_project_path)?;
-            fs::rename(tmp_apama_project_path, tedge_apama_project_path)?;
-            println!("Installation of new project successful");
+fn install_project(project_archive_path: &Path) -> Result<(), InternalError> {
+    let tedge_apama_project_path: &Path = Path::new(TEDGE_APAMA_PROJECT_DIR);
+    let tmp_apama_project_path: &Path = Path::new(TMP_APAMA_PROJECT_DIR);
 
-            println!("Restarting apama to load the new project");
-            run_cmd("service", "apama restart")?;
-            println!("Restart of apama service successful");
+    let archive_file = File::open(&project_archive_path)?;
 
-            success
-        }
+    let mut archive = zip::ZipArchive::new(archive_file)?;
 
-        PluginOp::Remove {
-            module: _,
-            version: _,
-        } => {
-            if tedge_apama_project_path.exists() {
-                println!("Stopping apama service");
-                run_cmd("service", "apama stop")?;
-                println!("Stopping apama service successful");
+    if let Err(zip::result::ZipError::FileNotFound) = archive.by_name("project/") {
+        return Err(InternalError::InvalidProjectArchive);
+    }
 
-                println!("Removing existing project at {}", TEDGE_APAMA_PROJECT_DIR);
-                fs::remove_dir_all(tedge_apama_project_path)?;
-                println!("Removal of existing project successful");
-            }
+    println!("Extracting the archive at {:?}", project_archive_path);
+    archive.extract(tmp_apama_project_path)?;
+    println!("Extraction successful");
+
+    // Deleting existing project as the rename API expects the target dir to be empty
+    delete_project()?;
+
+    println!(
+        "Installing newly extracted project to {}",
+        TEDGE_APAMA_PROJECT_DIR
+    );
+    fs::create_dir_all(tedge_apama_project_path)?;
+    fs::rename(
+        tmp_apama_project_path.join("project"),
+        tedge_apama_project_path,
+    )?;
+    println!("Installation of new project successful");
+
+    restart_apama_service()?;
+
+    Ok(())
+}
+
+fn restart_apama_service() -> Result<(), InternalError> {
+    println!("Restarting apama to load the new project");
+    run_cmd("service", "apama restart")?;
```

Comment:
What is this service command? Is it installed when you install apama service?

Reply:
That service file is provided by apama team here: https://github.com/thin-edge/thin-edge.io_examples/blob/main/StreamingAnalytics/src/service/apama

It's not part of the standard Apama community installation though. These service definitions are maintained separately.